### ISSUE TITLE
Log entries associated to wrong logName

### ISSUE THREAD

**Issue by protyposis** [googleapis/nodejs-logging#943]
Hi,

I have a system where I upload bunches of logs in parallel. I am basically following the [example from the README](https://github.com/googleapis/nodejs-logging#using-the-client-library) and use one shared `Logging` instance with multiple `log = logging.log(logName)` `Log` instances with different `logNames`.

I am getting cases where overlapping `log.write(manyLogEntries)` lead to entries uploaded with `Log` instance named `log-A` ending up in the logs of `Log` instance `log-B`. For validation, I am additionally adding the `logName` to `metadata.labels`. The resulting log entries correctly carry the metadata label `log-A`, but their `logName` is `log-B`.

A and B are just examples, it's actually many more instances and the misassociations seem to happen "randomly" between the instances. I was unsuccessful in reproducing this with multiple approaches outside my production system, even though the same code is used. It might be related to the large volume of logs being handled by the production system.

Is anybody aware of such a problem? Are multiple parallel `Log` instances supported?

**Comment by stephenplusplus** [comment]
@protyposis thank you for reporting. I have been testing this rather extensively, but haven't been able to reproduce. Here's a script I put together to try to create a level of chaos to introduce the error. (Note, it's required to update to @google-cloud/logging v9, released today).

Is there anything else you can think of to add?

```js
const assert = require('assert');
const Logging = require('@google-cloud/logging').Logging;
const v4 = require('uuid').v4;

const TESTS_PREFIX = 'issue-943';
const TIME_TEST_STARTED = new Date().toISOString();

const NUM_LOGS_TO_CREATE = 50;
const NUM_ENTRIES_TO_WRITE = 100000;
const FLUSH_AFTER_NUM_LOGS_FILLED = NUM_LOGS_TO_CREATE / 10;
const NUM_WRITES_TO_MAKE = 50;
const WRITE_CONSISTENCY_DELAY_MS = 10000;

const logging = new Logging();

function generateName() {
  return `${TESTS_PREFIX}-${Date.now()}-${v4().split('-').pop()}`;
}

run();

async function run() {
  const logNamesCreated = [];

  function validateEntriesFromLogCallback(log, callback) {
    let numEntriesReceived = 0;

    setTimeout(pollForMessages, WRITE_CONSISTENCY_DELAY_MS, {
      autoPaginate: false,
      filter: `timestamp > "${TIME_TEST_STARTED}"`,
      pageSize: 1000,
    });

    function pollForMessages(query) {
      log.getEntries(query, (err, entries, nextQuery) => {
        if (err) {
          callback(err);
          return;
        }

        assert(entries.every(e => e.data.originLog === log.name));

        numEntriesReceived += entries.length;
        console.log(`${numEntriesReceived}/${NUM_ENTRIES_TO_WRITE} entries received`);

        if (nextQuery) {
          setTimeout(pollForMessages, WRITE_CONSISTENCY_DELAY_MS, nextQuery);
          return;
        } else {
          if (numEntriesReceived < NUM_ENTRIES_TO_WRITE) {
            callback(new Error(`Not enough entries received`));
            return;
          }
        }

        callback();
      });
    }
  }

  function validateEntriesFromLog(log) {
    return new Promise((res, rej) => {
      validateEntriesFromLogCallback(log, (err, entries) => {
        if (err) {
          rej(err);
          return;
        }
        res(entries);
      });
    });
  }

  const logAndEntrySets = [];

  let numLogsCreated = 0;
  while (numLogsCreated++ < NUM_LOGS_TO_CREATE) {
    const logName = generateName();
    logNamesCreated.push(logName);
    const log = logging.log(logName);
    log.shortName = `log #${numLogsCreated}`;
    logAndEntrySets.push({log, entries: []});
  }

  let numLogsFilled = 0;
  while (numLogsFilled < NUM_LOGS_TO_CREATE) {
    const logAndEntrySetBatch = logAndEntrySets.splice(0, FLUSH_AFTER_NUM_LOGS_FILLED);

    for (const {log, entries} of logAndEntrySetBatch) {
      let numEntriesWritten = 0;
      while (numEntriesWritten++ < NUM_ENTRIES_TO_WRITE) {
        entries.push(log.entry({originLog: log.name}));
      }
    }

    console.log('Flushing logs...');
    await writeLogs(logAndEntrySetBatch);
    numLogsFilled += logAndEntrySetBatch.length;
  }

  async function writeLogs(logAndEntrySets) {
    const numEntriesPerWrite = NUM_ENTRIES_TO_WRITE / NUM_WRITES_TO_MAKE;
    let numWrites = 0;

    while (numWrites < NUM_WRITES_TO_MAKE) {
      const numEntriesWritten = numWrites * numEntriesPerWrite;
      for (const {log, entries} of logAndEntrySets) {
        console.log(
          `${log.shortName} writing ${numEntriesWritten}-${
            numEntriesWritten + numEntriesPerWrite
          }/${NUM_ENTRIES_TO_WRITE} entries`
        );
        await log.write(entries.splice(0, numEntriesPerWrite));
      }
      numWrites++;
    }
  }

  console.log('Validating logs...');
  let logNamesProcessed = 0;
  for (const logName of logNamesCreated) {
    console.log(`Validating log #${++logNamesProcessed}: ${logName}`);
    const log = logging.log(logName);
    await validateEntriesFromLog(log);
  }
}
```

**Comment by protyposis** [comment]
Thank you for looking into this. I have also attempted to create multiple different repro cases and wasn't able to reproduce artificially, while still being able to reproduce in the production system. Your script does about the same what I did.

Unfortunately I failed to record the exact library version (and the versions of its dependencies) with which we have observed this issue. It was definitely within the range of 8.0.1 - 8.0.9. I tried my repro cases with all these versions and couldn't reproduce with any.

Interestingly, since we updated to 8.1.1 due to issue #947, we weren't able to reproduce this any more. It might be a coincidence because it only happened "randomly" anyway, but there's a chance that it is fixed. Are you aware of any changes in this library or it's dependencies that might have fixes this?

Since it seems our way forward is unclear, and given the chance that this is fixed, I suggest we close this issue and I'll keep monitoring the situation and reopen if it happens again?
